src/net/posix-stack.cc

```diff
@@ -525,10 +526,114 @@ get_port_or_counter(const socket_address& sa) {
     }
 }
 
+static
+proxy_protocol_v2_header
+local_proxy_protocol_v2_header(const pollable_fd& pfd) {
+    auto& fd = pfd.get_file_desc();
+
+    auto local_sa = fd.get_address();
+    auto remote_sa = fd.get_remote_address();
+
+    proxy_protocol_v2_header header = {
+        .remote_address = remote_sa,
+        .local_address = local_sa,
+    };
+
+    return header;
+}
+
+// Parses proxy protocol v2 header; returns std::nullopt if no valid header is found.
+static
+future<std::optional<proxy_protocol_v2_header>>
+read_proxy_protocol_v2_header(pollable_fd& fd) {
+    constexpr size_t pp2_header_len = 16;
+    char header_buf[pp2_header_len];
+    auto n_read = co_await fd.read_some(header_buf, pp2_header_len);
+    if (n_read < pp2_header_len) {
+        co_return std::nullopt;
+    }
+    static const char pp2_signature[12] = {
+        0x0d, 0x0a, 0x0d, 0x0a, 0x00, 0x0d, 0x0a, 0x51, 0x55, 0x49, 0x54, 0x0a
+    };
+    if (std::memcmp(header_buf, pp2_signature, sizeof(pp2_signature)) != 0) {
+        co_return std::nullopt;
+    }
+
+    auto len = read_be<uint16_t>(header_buf + 14);
+
+    char stack_buffer[36]; // Suitable for IPv6 without extra TLVs
+    std::unique_ptr<char[]> heap_buffer;
+    auto* buffer = stack_buffer;
+
+    if (len > sizeof(stack_buffer)) {
+        heap_buffer = std::make_unique<char[]>(len);
+        buffer = heap_buffer.get();
+    }
+
+    auto xlen = co_await fd.read_some(buffer, len);
+    if (xlen < len) {
+        co_return std::nullopt;
+    }
+
+    uint8_t fam_proto = header_buf[13];
+    switch (header_buf[12]) { // version and command
+    case 0x20: // v2, LOCAL
+        if (fam_proto != 0x00) { // UNSPEC
+            co_return std::nullopt;
+        }
+        co_return local_proxy_protocol_v2_header(fd);
+    case 0x21: // v2, PROXY
+        // Mainline continues after the switch
+        break;
+    default:   // Not defined, must reject
+        co_return std::nullopt;
+    }
+
+    auto fam = fam_proto >> 4;
+    auto proto = fam_proto & 0x0F;
+
+    if (proto != 0x1) { // STREAM
+        co_return std::nullopt;
```

Comment:
This return nullopt is different from the others - while the previous ones were cases where the connection was genuinely broken and should be dropped, the case here includes cases (like UDP) which *are* supported by the protocol, but you chose not to support right now. This is fine, but we should document which pieces we don't support.